OSDOCS-3809 initial draft of overview for hosted control planes (hypershift) (#47565)
openshift/openshift-docs: modules/hosted-control-planes-overview.adoc

// Module included in the following assemblies:
//
// * architecture/control-plane.adoc

:_content-type: CONCEPT
[id="hosted-control-planes-overview_{context}"]
= Overview of hosted control planes (Technology Preview)

You can use hosted control planes for Red Hat {product-title} to reduce management costs, optimize cluster deployment time, and separate management and workload concerns so that you can focus on your applications.

You can enable hosted control planes as a Technology Preview feature through the link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.5/html/multicluster_engine/requirements-and-recommendations[multicluster engine for Kubernetes operator version 2.0] on Amazon Web Services (AWS).

:FeatureName: Hosted control planes
include::snippets/technology-preview.adoc[]

> == Architecture of hosted control planes

Reviewer: Needs an ID above it

While the coupled model works well, some situations require an architecture where the control plane and data plane are decoupled. In those cases, the data plane is on a separate network domain with a dedicated physical hosting environment. The control plane is hosted by using high-level primitives such as deployments and stateful sets that are native to Kubernetes. That is, the control plane is treated as any other workload.

image::hosted-control-planes-diagram.png[Diagram that compares the hosted control plane model against OpenShift with a coupled control plane and workers]

> == Benefits of hosted control planes

Reviewer: This needs an ID above it

Author: @bergerhoffer Like this?

```
[id="hosted-control-planes-benefits_{context}"]
== Benefits of hosted control planes
```

With hosted control planes for {product-title}, you can pave the way for a true hybrid-cloud approach and enjoy several other benefits.

* The security boundaries between management and workloads are stronger because the control plane is decoupled and hosted on a dedicated hosting service cluster. As a result, you are less likely to leak credentials for clusters to other users. Because infrastructure secret account management is also decoupled, cluster infrastructure administrators cannot accidentally delete control plane infrastructure.

* With hosted control planes, you can run many control planes on fewer nodes. As a result, clusters are more affordable.

> * Because the control planes consist of pods that are launched on {product-title}, control planes start quickly. In addition, because the control plane is another Kubernetes workload, the same principles apply, such as monitoring, logging, and auto-scaling.

Reviewer 2: I feel like _in addition_ and _the same_ are doing the same job. And I think you have made the case for the CP running as a workload in the first sentence. WDYT of a bit of minimalism in the second sentence?
```suggestion
* Because the control planes consist of pods that are launched on {product-title}, control planes start quickly. The same principles apply to <processes/workloads> such as monitoring, logging, and auto-scaling.
```

//
// * architecture/control-plane.adoc

:_content-type: CONCEPT
[id="hosted-control-planes-overview_{context}"]
= Overview of hosted control planes (Technology Preview)

You can use hosted control planes for Red Hat {product-title} to reduce management costs, optimize cluster deployment time, and separate management and workload concerns so that you can focus on your applications.

You can enable hosted control planes as a Technology Preview feature through the link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.5/html/multicluster_engine/requirements-and-recommendations[multicluster engine for Kubernetes operator version 2.0] on Amazon Web Services (AWS).

:FeatureName: Hosted control planes
include::snippets/technology-preview.adoc[]

== Architecture of hosted control planes

Typically in {product-title}, a cluster consists of a control plane and a data plane. The control plane includes an API endpoint, a storage endpoint, a workload scheduler, and an actuator that ensures state. The data plane is the hot path, including compute, storage, and networking where workloads and applications run.

> The control plane and the data plane are coupled in terms of locality. The control plane is hosted by a dedicated group of nodes, which can be physical or virtual, with a minimum number to ensure quorum. The network stack is shared, and admin access to a cluster offers visibility into the cluster's control plane, machine management APIs, and other system components that contribute to the state of a cluster.

Reviewer 3: May be clarify this paragraph is still describing the standalone model.

While the coupled model works well, some situations require an architecture where the control plane and data plane are decoupled. In those cases, the data plane is on a separate network domain with a dedicated physical hosting environment. The control plane is hosted by using high-level primitives such as deployments and stateful sets that are native to Kubernetes. That is, the control plane is treated as any other workload.

image::hosted-control-planes-diagram.png[Diagram that compares the hosted control plane model against OpenShift with a coupled control plane and workers]

== Benefits of hosted control planes

With hosted control planes for {product-title}, you can pave the way for a true hybrid-cloud approach and enjoy several other benefits.

* The security boundaries between management and workloads are stronger because the control plane is decoupled and hosted on a dedicated hosting service cluster. As a result, you are less likely to leak credentials for clusters to other users. Because infrastructure secret account management is also decoupled, cluster infrastructure administrators cannot accidentally delete control plane infrastructure.

> * With hosted control planes, you can run many control planes on fewer nodes. As a result, high availability (HA) clusters are more affordable.

Reviewer 4: ```suggestion
* With hosted control planes, you can run many control planes on fewer nodes. As a result, clusters are more affordable.
```

Author: I'll make that update and then ask Alberto and Jie for their review/approval. Thank you!

// Module included in the following assemblies:
//
// * architecture/control-plane.adoc

:_content-type: CONCEPT
[id="hosted-control-planes-overview_{context}"]
= Hosted control planes overview (technology preview)

> Hosted control planes for {product-title} provides a way to decouple your control plane from your worker nodes so that you can work in a hybrid cloud approach. You can enable hosted control planes through the multicluster engine for Kubernetes operator version 2.0. In that version, only Amazon Web Services (AWS) hosted clusters are available.

Reviewer 4: Hosted control planes helps Red Hat OpenShift customers focus on applications by reducing management overhead, optimizing cluster deployment time, and enabling hard separation of concerns between management and workloads.

 Hosted control planes is available on Amazon Web Services (AWS) as a [technology preview](https://docs.openshift.com/container-platform/4.10/rest_api/understanding-compatibility-guidelines.html#api-compatibility-common-terminology-dev-preview_compatibility-guidelines) feature enabled through the [multicluster engine for Kubernetes operator](https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.5/html/multicluster_engine/requirements-and-recommendations) version 2.0.

> :FeatureName: Hosted control planes

Reviewer 4: It's actually "hosted control planes" all lower-case but when starting a sentence it gets tricky :)

Author: I can try to rewrite any sentences that begin with "hosted control planes".

Reviewer: You could say "Support for hosted control planes"

include::snippets/technology-preview.adoc[]

== Hosted control planes architecture

Typically in {product-title}, a cluster consists of a control plane and a data plane. The control plane includes an API endpoint, a storage endpoint, a workload scheduler, and an actuator that ensures state. The data plane is the hot path, including compute, storage, and networking where workloads and applications run.

The control plane and the data plane are coupled in terms of locality. The control plane is hosted by a dedicated group of nodes, which can be physical or virtual, with a minimum number to ensure quorum. The network stack is shared, and admin access to a cluster offers visibility into the cluster's control plane, machine management APIs, and other system components that contribute to the state of a cluster.

> While the coupled model works well, some situations require an architectural model where the control plane and data plane ar decoupled. In the decoupled model, the data plane is on a separate network domain with a dedicated physical hosting environment. The control plane is hosted by using high-level primitives such as deployments and stateful sets that are native to Kubernetes.

Reviewer 4: ```suggestion
While the coupled model works well, some situations require an architectural model where the control plane and data plane are decoupled. In the decoupled model, the data plane is on a separate network domain with a dedicated physical hosting environment. The control plane is hosted by using high-level primitives such as deployments and stateful sets that are native to Kubernetes. I.e., the control-plane is treated as any other workload.
```

image::hosted-control-planes-diagram.png[Diagram that compares the hosted control plane model against OpenShift with a coupled control plane and workers]

== Benefits of hosted control planes

> Hosted control planes for {product-title} paves the way for a hybrid-cloud approach and offers several other benefits.

Reviewer 4: ```suggestion
Hosted control planes for {product-title} paves the way for a true hybrid-cloud approach and offers several other benefits.
```

> The management plane for control planes and cloud credentials is separate from the cluster that you interact with. Because the control plane is managed, you are less likely to accidentally delete your clusters.

Reviewer 4: Do we want to bullet the benefits?

Reviewer 4: By decoupling the control-plane and hosting them away on a dedicated hosting service cluster, hosted control planes hardens the security boundaries between management and workloads. For example, it is less likely to leak credentials to other users for the clusters. 

Furthermore, the decoupling infrastructure secret account management, as a result, as cluster infrastructure admin it is not possible to accidentally delete control-plane infrastructure.

> When your control plane and data plane are coupled, you might host one control plane on three machines. With hosted control planes, you can host several control planes on three machines, and you can run hundreds of control planes on 150 nodes. As a result, high availability (HA) clusters are more affordable.

Reviewer 4: It is possible to flatten out a cluster's fixed cost by re-using the hosting service cluster to host multiple control-plane, i.e., control-planes can be bin-packed to reduce compute overhead. This can results in at least threefold decrease in management overhead for provisioning OpenShift clusters.

Reviewer 4: added a comment on the newer review, the sentence contains duplicate context

> Because the control plane consists of pods that are launched on {product-title}, you do not need to wait to provision machines. In addition, you gain Kubernetes features:

Reviewer 4: ```suggestion
Because the control plane consists of pods that are launched on {product-title}, control-plane bootstrap time is minimal. In addition, because the control-plane is just another Kubernetes workload, the same principles apply (e.g., monitoring, logging, autoscaling, etc).
```

Reviewer 4: the problem with less or faster is that it compares, I try to avoid comparing if I can but It should workas well as it's a fact :)

Author: I rewrote the phrase to avoid making a comparison:

>Because the control planes consist of pods that are launched on OpenShift Container Platform, control planes start quickly.

   * Horizontal Pod Autoscaler (HPA)
   * Vertical Pod Autoscaler (VPA)
   * HA in the form of replicas

>    * Control plane hibernation

Reviewer 4: We can remove these as not all are available during the first release.

From an infrastructure perspective, you can push registries, HAProxy, cluster monitoring, storage nodes, and other infrastructure components to the tenant's cloud provider account, isolating usage to the tenant.

You can upgrade any consolidated control planes out of cycle from the segmented worker nodes, including embargoed Common Vulnerabilities and Exposures (CVEs).

> If needed, you can run control planes on one CPU chip type, such as x86, and run worker nodes on another type, such as ARM or IBM Power.

Reviewer 4: Future benefit, we can ommit today
